src/data/repositories/DataSetD2Api.ts

```diff
@@ -1,91 +1,102 @@
     }
 
     get(options: D2DataSetOptions): FutureData<Paginated<DataSet>> {
-        return this.getBaseData({ projectsIds: options.filters.projectsIds }).flatMap(
-            ({ coreCompetencies }) => {
-                const attributesToFilter = options.filters.projectsIds
-                    ? [this.metadata.attributes.project.id]
-                    : [this.metadata.attributes.createdByApp.id];
-                return apiToFuture(
-                    this.api.models.dataSets.get({
-                        pageSize: options.paging.pageSize,
-                        page: options.paging.page,
-                        filter: {
-                            "attributeValues.attribute.id": { in: attributesToFilter },
-                            "attributeValues.value": this.isProjectFilterPresent(options)
-                                ? { in: [...this.getOrThrow(options.filters.projectsIds)] }
-                                : { eq: "true" },
-                            identifiable: { token: options.filters.search },
-                            id: { in: options.filters.ids },
-                        },
-                        fields: {
-                            ...dataSetFields,
-                            ...(options.includeOrgUnits
-                                ? {
-                                      organisationUnits: {
-                                          id: true,
-                                          path: true,
-                                          displayName: true,
-                                      },
-                                  }
-                                : {}),
-                        },
-                        order: `${options.sorting.field}:${options.sorting.order}`,
-                    })
-                ).flatMap(d2Response => {
-                    const projectIds = this.getProjectIds(d2Response.objects);
-                    return this.getProjectsByIds(projectIds).map(projects => {
-                        const dataSets = d2Response.objects.map(d2DataSet => {
-                            return this.buildDataSet(d2DataSet, coreCompetencies, projects);
-                        });
-                        return { ...d2Response.pager, data: dataSets };
+        return this.getBaseData().flatMap(({ attributes, coreCompetencies }) => {
+            const attributesToFilter = options.filters.projectsIds
+                ? [attributes.project.id]
+                : [attributes.createdByApp.id];
+            return apiToFuture(
+                this.api.models.dataSets.get({
+                    pageSize: options.paging.pageSize,
+                    page: options.paging.page,
+                    filter: {
+                        "attributeValues.attribute.id": { in: attributesToFilter },
+                        "attributeValues.value": this.isProjectFilterPresent(options)
+                            ? { in: [...this.getOrThrow(options.filters.projectsIds)] }
```

Comment:
I get the idea, we're trying to be declarative on the conditional. The problem is that TS is then unaware of the check and you have to add the hacky getOrThrow. One simple solution is write the direction condition: `"attributeValues.value": options.filters.projectsIds`. Whenever the predicate is really complex and we need a function, we can check how to do it.